text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,330 @@
+
+## Datamodel Changes
+
+Given our desired design and algorithms for detecting, merging and manipulating Entities, we need the ability to denote how entity and resource relate. These changes must not break existing usage of Resource, therefore:
+
+- The Entity model must be *layered on top of* the Resource model.  A system does not need to interact with entities for correct behavior.
+- Existing key usage of Resource must remain when using Entities, specifically navigationality (see: [OpenTelemetry Resources: Principles and Characteristics](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit))
+- Downstream components should be able to engage with the Entity model in Resource.
+
+The following changes are made:
+
+### Resource
+
+| Field | Type | Description | Changes |
+| ----- | ---- | ----------- | ------- |
+| schema_url | string | The Schema URL, if known. This is the identifier of the Schema that the resource data  is recorded in.  This field is deprecated and should no longer be used. | Will be deprecated |
+| dropped_attributes_count |  integer | dropped_attributes_count is the number of dropped attributes. If the value is 0, then no attributes were dropped. | Unchanged |
+| attributes | repeated KeyValue | Set of attributes that describe the resource.<br/><br/>Attribute keys MUST be unique (it is not allowed to have more than one attribute with the same key).| Unchanged |
+| entities | repeated ResourceEntityRef | Set of entities that participate in this Resource. | Added |
+
+The DataModel would ensure that attributes in Resource are produced from both the identifying and descriptive attributes of Entity.  This does not mean the protocol needs to transmit duplicate data, that design is TBD.
+
+### ResourceEntityRef
+
+The entityref data model, would have the following changes from the original [entity OTEP](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md) to denote references within Resource:
+
+| Field | Type | Description | Changes |
+| ----- | ---- | ----------- | ------- |
+| schema_url | string | The Schema URL, if known. This is the identifier of the Schema that the entity data  is recorded in. To learn more about Schema URL ([see docs](https://opentelemetry.io/docs/specs/otel/schemas/#schema-url)) | added |
+| type | string | Defines the type of the entity. MUST not change during the lifetime of the entity. For example: "service" or "host". This field is required and MUST not be empty for valid entities. | unchanged |
+| identifying_attributes_keys | repeated string | Attribute Keys that identify the entity.<br/>MUST not change during the lifetime of the entity. The Id must contain at least one attribute.<br/><br/>These keys MUST exists in Resource.attributes.<br/><br/>Follows OpenTelemetry common attribute definition. SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |
```

Comment:
Is adding or removing a value from the set considered a change? Do entities have versions? If so, can entities be referred to from places where the version is not known?

Reply:
Yes - `schema_url` can be considered a version for telemetry within OTel -> this would be the version of the identity.

Yes - we likely need to deal with the possibility folks do not specify a schema_url for an entity. In this case, it would be an unknown version and we are unable to determine if we can merge.  I'm updating the merge algorithm definition shortly, so I'll account for that in the update.

Reply:
I wasn't referring to schema versions. I was thinking about something like `service.version` for entities to signal the version of the content, not the type.

Reply:
How would the version not be part of the entity identity in that case?

Reply:
OK, I spent some time reading through all the documents and some referenced documents. I like the overall approach, but fear that the lack of clarity might be an issue. Would it make sense to ask my organization if I could contribute to such projects and somehowe become a contributing member of this work stream? Something that I really do not get (and that's maybe the source of my confusion) is why we would send the attribute names twice. If the proposal envisions separate entity state events (which is something I favor), it would make sense to reduce the data sent with data points as much as possible. Also, not sending the attribute keys here would completely remove situations where the set of identifying attributes varies. Furthermore, the MUST not change statements here are not correct. They CANNOT change as they define equivalence of the underlying entity and changing them must result in a separate entity instance. Finally, I still struggle with partial identification: The service entity, for compatibility reasons, would be identified by `{service.name, service.namespace, service.instance.id}`. However, in practice, there are cases where e.g. `service.instance.id` isn't known when producing the data. It might get added later by a processor or it might stay like that. Using entities might fix that, but the transition towards entities would result in a phase where this issue exists and must be dealt with.

Reply:
Sorry for not responding to this earlier.

IF you're interested in contributing, we are trying to organize our prototyping work.  You can join/discuss with us live (or watch recordings of) the [Specification: Entities WG meeting](https://github.com/open-telemetry/community/?tab=readme-ov-file#specification-sigs).

Regarding sending keys twice:  The rationale is in the document, but to explicitly state it:

- We cannot break existing `Resource` usage, which means the `attributes` fields MUST be usable without understanding or interacting with the new entity model. This means *all* attributes from entities MUST be in this portion of the proto or we've broken open telemetry.
- We debated between the following options (called out in options considered):
   - Duplicate attributes in `Resource.attributes` and `Resource.entities`.
   - Reference `Resource.attributes` in `Resource.entitites` by index
   - Reference `Resource.attributes` in `Resource.entities` by name
   
   We chose the last option for the proposal, but this may change as we prototype and move forward.  The first requirement will not.

Regarding an `entity` where identity is not known when adding to resource: Going forward we're basically outlawing that use case.  For service this means, you'll need to attach service name and instance when attaching service.  We have done work in semconv and the specification such that we can attach service.instance.id when producing data.  Additioanlly, if you want a different instance id you can do one of two things:

1. Update instance id in the collector via an entity-override
2. Create an association between the "service reported by the sdk" and the "service you get identity from somewhere else".

Reply:
I'll try to join the WG on behalf of my organization or my employer.

Until then: Could we consider not setting any reference in the gRPC messages and having them behind the schema_url as a fourth option at least? I agree that not breaking existing `attributes` and having entities as transparent extension are non-negotiable requirements. Regarding the instance.id discussion: I think it makes sense to explain why I'm raising this example once I figured under which mandate I can join the WG.

Reply:
(I think we can resolve this discussion by adding the option because the other parts should not block progress)

Reply:
The mechanism of how to understand entity semantics in Resource is an open question below, meaning while this OTEP proposing something, we aren't committed until we have prototypes we're comfortable with and these would enter the specification.

While I discuss in a different comment why using the contents of schema_url is off the table for the general algorithm, it CAN be used when we have it. Look forward to the ideas you have here.

Marking this resolve, hope you're able to join.